File: src/staticTyping/structural/ocaml/duck.ml

(*type duck = <quake : string ; dance : string>*)

(* A virtual class instead of a type to make mallard an explicit sub-class of duck *)
class virtual duck = object
	method virtual quake : string
	method virtual dance : string
end

class mallard = object
	inherit duck
	method quake : string = "quack quack"
	method dance : string = " _/¯ "
end

class wolf = object
	method quake : string = "QUACK QUACK WHOO"
	method dance : string = " ¯\\_()_/¯ "
	method eat(d : duck) : string = " 😈 "
end

let thewolf : wolf = new wolf;;
let theDuck : duck = (thewolf :> duck);;
let aDuck : duck = new mallard;;


Printf.printf "%s\n" aDuck#quake;
Printf.printf "%s\n" theDuck#quake;

Printf.printf "%s\n" aDuck#dance;
Printf.printf "%s\n" theDuck#dance;

Printf.printf "%s\n" (thewolf#eat(aDuck));

File: src/staticTyping/structural/go/duck.go
package main

import "fmt"

type Duck interface {
	quack() string
	dance() string
}

type Wolf struct {
}

type Mallard struct {
}


func (mallard Mallard) quack() string { return "quack quack" }
func (mallard Mallard) dance() string  { return " _/¯ " }

func (wolf Wolf) quack() string { return "QUACK QUACK WHOO" }
func (wolf Wolf) dance() string  { return " ¯\\_()_/¯ " }
func (wolf Wolf) eat(duck Duck) string  { return " 😈 " }


func main() {
	var wolf Wolf
	var theDuck Duck

	theDuck = &wolf /* Type inference between Wolf and Duck. & used not to copy the wolf but to use its reference. */

	var aDuck Mallard

	twoDucksAlone(aDuck, theDuck)

	fmt.Printf("%s\n", wolf.eat(aDuck))
}

func twoDucksAlone(d1 Duck, d2 Duck) {
	fmt.Printf("%s\n", d1.quack())
	fmt.Printf("%s\n", d2.quack())

	fmt.Printf("%s\n", d1.dance())
	fmt.Printf("%s\n", d2.dance())
}
